### PR TITLE
esp32: Fix PWM not allowing frequencies < 611 Hz.

### PR DESCRIPTION
**Pull request by robert-hh**
Tested with duty_u16(), duty() and duty_ns() for frequencies down to
1 Hz.
Fixes issue #8189

### CODE REVIEW

File: ports/esp32/machine_pwm.c [hunk 1]
```diff
@@ -296,11 +295,18 @@ STATIC int duty_to_ns(machine_pwm_obj_t *self, int duty) {
...
+        duty <<= HIGHEST_PWM_RES + UI_RES_SHIFT - resolution;
+    } else {
+        duty >>= -(HIGHEST_PWM_RES + UI_RES_SHIFT - resolution);
+    }
+    return duty;
 }
 
 STATIC uint32_t get_duty_u10(machine_pwm_obj_t *self) {
-    return get_duty_u16(self) >> (HIGHEST_PWM_RES - PWRES);
+    return get_duty_u16(self) >> 6;
```
Comment (dpgeorge) [inline ports/esp32/machine_pwm.c:309]: The the "6"?  Probably worth a comment, or a #define constant.

Comment (robert-hh) [reply ports/esp32/machine_pwm.c:309]: That#s the difference between 16 bit of duty_u16() and 10 bit of duty(). But I can make it more clear.

**Comment by robert-hh** [comment]
Changed, rebased & squashed. The names of #defines chosen by the initial implementation could be more distinctive, but left them as they were.

File: ports/esp32/machine_pwm.c (hunk 1 above)
Comment (dpgeorge) [reply ports/esp32/machine_pwm.c]: Yes of course, I didn't see which function this line was in!  It is pretty clear when you see the u10 vs the u16...

**Comment by IhorNehrutsa** [comment]
Bug is here:
https://github.com/micropython/micropython/blob/30b6ce86be384ac04566d8f7bbf67d5906d6f307/ports/esp32/machine_pwm.c#L87-L92

Algorithm is developed in assumption that **HIGHEST_PWM_RES must be 16 or 14**
but LEDC_TIMER_BIT_MAX and  LEDC_TIMER_16_BIT is enum (not a macro) and we got
HIGHEST_PWM_RES == 20

I tried to fix with SOC_LEDC_TIMER_BIT_WIDE_NUM, but SOC_LEDC_TIMER_BIT_WIDE_NUM also is not difined in Micropython compile time

This is right version:
```
// Possible highest resolution in device
#if CONFIG_IDF_TARGET_ESP32
#define HIGHEST_PWM_RES (LEDC_TIMER_16_BIT) // 20 bit for ESP32, but 16 bit is used
#else
#define HIGHEST_PWM_RES (LEDC_TIMER_BIT_MAX - 1)
#endif
```
and then lines 
https://github.com/micropython/micropython/blob/30b6ce86be384ac04566d8f7bbf67d5906d6f307/ports/esp32/machine_pwm.c#L299
https://github.com/micropython/micropython/blob/30b6ce86be384ac04566d8f7bbf67d5906d6f307/ports/esp32/machine_pwm.c#L315
are right too.

**Comment by robert-hh** [comment]
With:
`#define UI_RES_SHIFT (UI_RES_16_BIT - HIGHEST_PWM_RES)`
the expression:
`HIGHEST_PWM_RES + UI_RES_SHIFT - timer.duty_resolution`
results in:
`UI_RES_16_BIT - timer.duty_resolution`
So care must be taken about duty_resolution being larger or smaller than 16.

**Comment by IhorNehrutsa** [comment]
@robert-hh 
Yes, you are right, there are two ways. May try to compare both.

**Comment by robert-hh** [comment]
Tested with your version works fine as well on a generic ESP32. The Frequency deviation of 1001.6 Hz instead of 1000Hz  stay. GENERIC_C3 and UM_FEATHERS2 show better frequency resolution and granularity.

The previous code worked also well on a UM_FEATHERS2 and a GENERIC_C3.
At the GENERIC_C3 down to 10Hz, at the UM_FEATHERS2 reliable down to 2 Hz, with a strange hiccup when changing from 10Hz -> 1Hz, but not when changing from 2Hz -> 1Hz.

File: ports/esp32/machine_pwm.c (hunk 1 above)
Comment (robert-hh) [reply ports/esp32/machine_pwm.c]: Changed back to '6' and a comment. That looks more obvious than the macro names.

**Comment by robert-hh** [comment]
In favor of @IhorNehrutsa's PR, being the initial contributor for this PWM module.

**Pull request closed without merging**